File: packages/control-contract/src/registration.ts
import type { TypedId } from "./ids.js";

export const CONTROL_PLANE_PROTOCOL_MAJOR = 1;

export type HostCapabilities = { readonly values: readonly string[] };

export type HostRegistration = {
  readonly agent_instance_id: TypedId;
  readonly workspaces: readonly TypedId[];
  readonly capabilities: HostCapabilities;
  readonly protocol_major: number;
};

/** Consumers must redact grant_token from logs and errors. */
export type HostRegistrationRequest = {
  readonly grant_token: string;
  readonly host: HostRegistration;
};

export type RegisteredHost = {
  readonly agent_instance_id: TypedId;
  readonly workspaces: readonly TypedId[];
  readonly capabilities: HostCapabilities;
  readonly registered_at_unix_seconds: number;
};

/** Non-secret control-plane readiness projection. */
export type ControlPlaneHealth = {
  readonly service_version: string;
  readonly protocol_major: number;
  readonly store_kind: string;
  readonly registered_host_count: number;
  readonly database_adapter_ready: boolean;
};

File: src-tauri/crates/altai-control-protocol/src/registration.rs
//! Versioned control-plane health and execution-host registration contracts.
//!
//! The registration grant is deliberately a plain request field. A transport
//! can accept it without retaining it; callers must redact the field in logs
//! and errors.

use crate::{AgentInstanceId, WorkspaceId};
use serde::{Deserialize, Serialize};
use std::collections::BTreeSet;

pub const CONTROL_PLANE_PROTOCOL_MAJOR: u16 = 1;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Default)]
pub struct HostCapabilities {
    pub values: BTreeSet<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct HostRegistration {
    pub agent_instance_id: AgentInstanceId,
    pub workspaces: Vec<WorkspaceId>,
    pub capabilities: HostCapabilities,
    pub protocol_major: u16,
}

/// Transport request; deliberately does not implement `Debug`.
#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct HostRegistrationRequest {
    pub grant_token: String,
    pub host: HostRegistration,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RegisteredHost {
    pub agent_instance_id: AgentInstanceId,
    pub workspaces: Vec<WorkspaceId>,
    pub capabilities: HostCapabilities,
    pub registered_at_unix_seconds: u64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ControlPlaneHealth {
    pub service_version: String,
    pub protocol_major: u16,
    pub store_kind: String,
    pub registered_host_count: usize,
    pub database_adapter_ready: bool,
}
